core/modules/replica/IngestClient.cc

```diff
@@ -0,0 +1,310 @@
+    _assertErrorCode(ec, __func__, "handshake send");
+
+    // Read and analyze the response
+    ProtocolIngestResponse response;
+    _readResponse(response);
+    if (response.status() != ProtocolIngestResponse::READY_TO_READ_DATA) {
+        _abort(__func__, "handshake receive, server error: " + response.error());
+    }
+    _numRowsPerSend = response.max_rows();
+    LOGS(_log, LOG_LVL_DEBUG, _context(__func__) << "_numRowsPerSend: " << _numRowsPerSend);
+}
+
+
+void IngestClient::_readResponse(ProtocolIngestResponse& response) {
+
+    LOGS(_log, LOG_LVL_DEBUG, _context(__func__));
+
+    // Start with receiving the fixed length frame carrying
+    // the size (in bytes) the length of the subsequent message.
+    const size_t frameLengthBytes = sizeof(uint32_t);
+
+    _bufferPtr->resize(frameLengthBytes);
+
+    boost::system::error_code ec;
+    boost::asio::read(
+        _socket,
+        boost::asio::buffer(_bufferPtr->data(), frameLengthBytes),
+        boost::asio::transfer_at_least(frameLengthBytes),
+        ec
+    );
+    _assertErrorCode(ec, __func__, "frame receive");
+
+    // Get the length of the message and try reading the message itself
+    // from the socket.
+    const uint32_t responseLengthBytes = _bufferPtr->parseLength();
+
+    _bufferPtr->resize(responseLengthBytes);    // make sure the buffer has enough space to
+                                                // accommodate the data of the message.
+    boost::asio::read(
+        _socket,
+        boost::asio::buffer(_bufferPtr->data(), responseLengthBytes),
+        boost::asio::transfer_at_least(responseLengthBytes),
+        ec
+    );
+    _assertErrorCode(ec, __func__, "response receive");
+
+    // Parse and analyze the response
+    try {
+        _bufferPtr->parse(response, responseLengthBytes);
+    } catch (exception const& ex) {
+        _abort(__func__, "response processing failed: " + string(ex.what()));
+    }
+}
+
+
+void IngestClient::_assertErrorCode(boost::system::error_code const& ec,
+                                    string const& func,
+                                    string const& msg) {
+    LOGS(_log, LOG_LVL_DEBUG, _context(__func__));
+
+    if (ec.value() != 0) {
+        _abort(func, msg + ", error: " + ec.message());
+    }
+}
+
+
+void IngestClient::_abort(string const& func,
+                          string const& error) {
+    LOGS(_log, LOG_LVL_DEBUG, _context(__func__));
+    _closeConnection();
+    string const msg = _context(func) + error;
+    LOGS(_log, LOG_LVL_ERROR, msg);
+    throw IngestClientError(msg);
+}
+
+                
+void IngestClient::_closeConnection() {
+
+    LOGS(_log, LOG_LVL_DEBUG, _context(__func__));
+
+    // Always attempt to shutdown and close the socket. This code deliberately
+    // ignores any abnormal conditions should they happen during the operation.
+    boost::system::error_code ec;
+    _socket.shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
+    _socket.close(ec);
```

Comment:
I guess it's ok to shutdown and close a socket multiple times? (the docs were not much help on this)

Reply:
I think it is harmless. It might be useful to log failures that aren't related to the socket already being closed.